system/jlib/jsocket.cpp

```diff
@@ -466,6 +467,200 @@ extern jlib_decl bool queryKeepAlive(int &time, int &intvl, int &probes)
         return false;
 }
 
+static bool getAddressInfo(const char *name, unsigned *netaddr, bool okToLogErr);
+
+static CriticalSection queryDNSCS;
+
+class GetAddrInfoThread;
+static std::list<GetAddrInfoThread *> gaPtrList;
+
+class GetAddrInfoThread : public Thread
+{
+    char name[256];
+    Semaphore semait;
+    std::atomic<bool> started;
+    std::atomic<bool> ended;
+    unsigned netaddr[4];
+    bool result;
+
+public:
+    GetAddrInfoThread(const char *_name) : started(false), ended(false), result(false)
+    {
+        strncpy(name, _name, 255);
+        name[255] = '\0';
+        memset(netaddr, 0, sizeof(netaddr));
+    }
+
+    bool thrdHasEnded()
+    {
+        if (started)
+            return ended;
+        return false;
+    }
+
+    bool waitms(unsigned timeoutms, unsigned *resAddr)
+    {
+        bool ret = semait.wait(timeoutms);
+        if (ret && ended)
+        {
+            if (result)
+                memcpy(resAddr, netaddr, sizeof(netaddr));
+            if (join())
+                return true;
+        }
+        {
+            CriticalBlock block(queryDNSCS);
```

Comment:
Or a pthread_kill() here ???

Reply:
I think we could pthread_cancel() here to help stop thread asap but still would need to check for it / cleanup in the reaper thread.